
project bootable/recovery/
diff --git a/bootable/recovery/Android.mk b/bootable/recovery/Android.mk
index 661a103..b0eb139 100644
--- a/bootable/recovery/Android.mk
+++ b/bootable/recovery/Android.mk
@@ -67,7 +67,7 @@ BOARD_RECOVERY_CHAR_HEIGHT := $(shell echo $(BOARD_USE_CUSTOM_RECOVERY_FONT) | c
 
 LOCAL_CFLAGS += -DBOARD_RECOVERY_CHAR_WIDTH=$(BOARD_RECOVERY_CHAR_WIDTH) -DBOARD_RECOVERY_CHAR_HEIGHT=$(BOARD_RECOVERY_CHAR_HEIGHT)
 
-BOARD_RECOVERY_DEFINES := BOARD_RECOVERY_SWIPE BOARD_HAS_NO_SELECT_BUTTON BOARD_UMS_LUNFILE BOARD_RECOVERY_ALWAYS_WIPES BOARD_RECOVERY_HANDLES_MOUNT BOARD_TOUCH_RECOVERY RECOVERY_EXTEND_NANDROID_MENU TARGET_USE_CUSTOM_LUN_FILE_PATH TARGET_DEVICE TARGET_RECOVERY_FSTAB
+BOARD_RECOVERY_DEFINES := RECOVERY_CHARGEMODE BOARD_RECOVERY_SWIPE BOARD_HAS_NO_SELECT_BUTTON BOARD_UMS_LUNFILE BOARD_RECOVERY_ALWAYS_WIPES BOARD_RECOVERY_HANDLES_MOUNT BOARD_TOUCH_RECOVERY RECOVERY_EXTEND_NANDROID_MENU TARGET_USE_CUSTOM_LUN_FILE_PATH TARGET_DEVICE TARGET_RECOVERY_FSTAB
 
 $(foreach board_define,$(BOARD_RECOVERY_DEFINES), \
   $(if $($(board_define)), \
diff --git a/bootable/recovery/recovery.c b/bootable/recovery/recovery.c
index fcaf0a6..43852fb 100644
--- a/bootable/recovery/recovery.c
+++ b/bootable/recovery/recovery.c
@@ -35,6 +35,7 @@
 #include "common.h"
 #include "cutils/properties.h"
 #include "cutils/android_reboot.h"
+#include <sys/reboot.h>
 #include "install.h"
 #include "minui/minui.h"
 #include "minzip/DirUtil.h"
@@ -158,6 +159,52 @@ fopen_path(const char *path, const char *mode) {
     if (fp == NULL && path != COMMAND_FILE) LOGE("Can't open %s\n", path);
     return fp;
 }
+#ifdef RECOVERY_CHARGEMODE
+void handle_chargemode() {
+    const char* filename = "/proc/cmdline";
+    struct stat file_info;
+    if (0 != stat(filename, &file_info))
+        return;
+    int file_len = file_info.st_size;
+    char* file_data = (char*)malloc(file_len + 1);
+    FILE *file = fopen(filename, "rb");
+    if (file == NULL)
+        return;
+    fread(file_data, file_len, 1, file);
+    file_data[file_len] = '\0';
+    fclose(file);
+    if (strstr(file_data, "androidboot.mode=offmode_charging") != NULL)
+        reboot(RB_POWER_OFF);
+}
+
+int get_chargemode() {
+    char buffer[1024], *p, *q;
+    int len;
+    int fd = open("/proc/app_info",O_RDONLY);
+    int flag = 0;
+    LOGI("get_chargemode.\n");
+        if (fd < 0) {
+        LOGE("could not open /proc/app_info:%s", strerror(errno));
+        }
+        do {
+            len = read(fd,buffer,sizeof(buffer)); }
+        while (len == -1 && errno == EINTR);
+
+    if (len < 0) {
+        LOGE("could not read /proc/app_info:%s", strerror(errno));
+        close(fd);
+    }
+    close(fd);
+	char * charge;
+	charge = strstr(buffer, "charge_flag");
+    LOGI("=QIWU=Charge Flag: %c\n", charge[13]);
+	if((int)charge[13] == 49 ){	
+		LOGI("=QIWU=Charge Flag: %c", charge[13]);
+		flag = 1;
+    }
+	return flag;
+}
+#endif
 
 // close a file, log an error if the error indicator is set
 static void
@@ -990,6 +1037,11 @@ main(int argc, char **argv) {
         }
         return busybox_driver(argc, argv);
     }
+
+#ifdef RECOVERY_CHARGEMODE
+    handle_chargemode();
+#endif
+
     __system("/sbin/postrecoveryboot.sh");
 
     int is_user_initiated_recovery = 0;
@@ -1000,6 +1052,15 @@ main(int argc, char **argv) {
     freopen(TEMPORARY_LOG_FILE, "a", stderr); setbuf(stderr, NULL);
     printf("Starting recovery on %s\n", ctime(&start));
 
+#ifdef RECOVERY_CHARGEMODE
+    int flag;
+    flag = get_chargemode();
+
+    if(flag == 1){
+        __system("charge");
+    }else {
+#endif
+
     device_ui_init(&ui_parameters);
     ui_init();
     ui_print(EXPAND(RECOVERY_VERSION)"\n");
@@ -1180,7 +1241,10 @@ main(int argc, char **argv) {
 
     return EXIT_SUCCESS;
 }
-
+#ifdef RECOVERY_CHARGEMODE
+	return 0;
+}
+#endif
 int get_allow_toggle_display() {
     return allow_display_toggle;
 }
