
project packages/services/Telephony/
diff --git a/packages/services/Telephony/src/com/android/phone/PhoneGlobals.java b/packages/services/Telephony/src/com/android/phone/PhoneGlobals.java
index 40f838f..5156d53 100644
--- a/packages/services/Telephony/src/com/android/phone/PhoneGlobals.java
+++ b/packages/services/Telephony/src/com/android/phone/PhoneGlobals.java
@@ -74,6 +74,10 @@ import com.android.phone.WiredHeadsetManager.WiredHeadsetListener;
 import com.android.server.sip.SipService;
 import com.android.services.telephony.common.AudioMode;
 
+// Patch airplane
+import android.net.ConnectivityManager;
+//End Patch
+
 /**
  * Global state for the telephony subsystem when running in the primary
  * phone process.
@@ -1035,7 +1039,18 @@ public class PhoneGlobals extends ContextWrapper implements WiredHeadsetListener
             if (action.equals(Intent.ACTION_AIRPLANE_MODE_CHANGED)) {
                 boolean enabled = System.getInt(getContentResolver(),
                         System.AIRPLANE_MODE_ON, 0) == 0;
+                // Patch. Airplane
+                ConnectivityManager cm = (ConnectivityManager)getSystemService(Context.CONNECTIVITY_SERVICE);
+                // End patch
                 phone.setRadioPower(enabled);
+                // Patch. Airplane
+                if (enabled) {
+                    boolean dmode = cm.getMobileDataEnabled();
+                    cm.setMobileDataEnabled(false);
+                    cm.setMobileDataEnabled(true);
+                    cm.setMobileDataEnabled(dmode);
+                }
+                // End patch
             } else if (action.equals(TelephonyIntents.ACTION_ANY_DATA_CONNECTION_STATE_CHANGED)) {
                 if (VDBG) Log.d(LOG_TAG, "mReceiver: ACTION_ANY_DATA_CONNECTION_STATE_CHANGED");
                 if (VDBG) Log.d(LOG_TAG, "- state: " + intent.getStringExtra(PhoneConstants.STATE_KEY));
