
project frameworks/av/
diff --git a/frameworks/av/services/camera/libcameraservice/api1/CameraClient.cpp b/frameworks/av/services/camera/libcameraservice/api1/CameraClient.cpp
index afaf12a..ed1e3c7 100644
--- a/frameworks/av/services/camera/libcameraservice/api1/CameraClient.cpp
+++ b/frameworks/av/services/camera/libcameraservice/api1/CameraClient.cpp
@@ -1006,7 +1006,7 @@ void CameraClient::handleCompressedBurstPicture(const sp<IMemory>& mem) {
     // get called only once. When burst finishes this message will get automatically
     // disabled in the respective call for restarting the preview.
 
-    sp<ICameraClient> c = mCameraClient;
+    sp<ICameraClient> c = mRemoteCallback;
     mLock.unlock();
     if (c != 0) {
         c->dataCallback(CAMERA_MSG_COMPRESSED_IMAGE, mem, NULL);
